uswds/uswds
# USWDS - Input: Move input width classes out of usa-form package

# Summary
**Moved `.usa-input--[width]` and `.usa-input-group--[width]` classes out of the `usa-form` package.** 
These classes are now generated in the `usa-input` and `usa-input-prefix-suffix` packages and can be used without the `.usa-form` parent element.

> [!Note]
> This is an alternative solution to PR #5935 (Thanks @aduth!)

## Breaking change

This is not a breaking change.

## Related issue

Closes #5312

## Related pull requests

[Changelog PR](https://github.com/uswds/uswds-site/pull/2997)

## Preview link

- [Input component](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/al-input-width/?path=/story/components-form-inputs-text-input--input)
- [Input prefix/suffix component](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/al-input-width/?path=/story/components-form-inputs-input-prefix-or-suffix--prefix)
- [Mailing address pattern](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/al-input-width/?path=/story/patterns-forms--mailing-address)

## Problem statement

As a developer, I expect that if I include the package listed on a component's documentation (e.g. [Text Input](https://designsystem.digital.gov/components/text-input/)), all of the usage options documented for that component are available, so that I'm not confused by styles not existing despite being documented, or be forced to include more packages than I'd otherwise need for my optimized installation.

## Solution
- Moved `.usa-input--[width]` and `.usa-input-group--[width]` class defintions into the `usa-input` and `usa-input-prefix-suffix` package, respectively.
    - Created a new `$system-input-widths` map in the `uswds-core` package so that the values can be defined in a single location
    - Used the `$system-input-widths` map to generate `usa-input--[width]` and `usa-input-group--[width]` classes
- Removed the `max-width: none` declaration from `.usa-form .usa-input` and `.usa-form .usa-textarea` to prevent specificity overrides

## Testing and review
- Confirm the `.usa-input--[width]` and `.usa-input-group--[width]` classes match the CSS (class names and values) in `develop`
- Confirm `usa-input--[width]` and `usa-input-group--[width]` classes work as expected both _with_ AND _without_ a `.usa-form` parent on the following elements:
    - `usa-input`
    - `usa-textarea`
    - `usa-input-group`
    - Character count input (currently accepts the `usa-input--[width]` classes in `develop`)
    - Input mask input (currently accepts the `usa-input--[width]` classes in `develop`)
- Confirm no regressions as a result of reducing the specificity of `max-width: none` on `.usa-form` children:
    - Set `$theme-input-max-width` to a value less than mobile (like "card" or 15)
    - Open the local build of [mailing address pattern story](http://localhost:6006/?path=/story/patterns-forms--mailing-address) 
    - Confirm that all elements in the form, including the zip code input, are the same width as they are in [develop](https://federalist-3b6ba08e-0df4-44c9-ac73-6fc193b0e19c.sites.pages.cloud.gov/preview/uswds/uswds/develop/?path=/story/patterns-forms--mailing-address)
- Confirm that creating `$system-input-widths` in `uswds-core` makes sense and lives in the appropriate location
